src/drt/src/dr/FlexDR.cpp

```diff
@@ -1230,6 +1235,167 @@ void FlexDR::reportGuideCoverage()
+    // The region is of size 3x3 GCELLBOX with the via in the center GCELLBOX
+    auto origin = via->getOrigin();
+    auto block = getDesign()->getTopBlock();
+    auto gcell_idx = block->getGCellIdx(origin);
+    Rect region, tmp_box;
+    tmp_box = block->getGCellBox({gcell_idx.x() - 1, gcell_idx.y() - 1});
+    region.set_xlo(tmp_box.xMin());
+    region.set_ylo(tmp_box.yMin());
+    tmp_box = block->getGCellBox({gcell_idx.x() + 1, gcell_idx.y() + 1});
+    region.set_xhi(tmp_box.xMax());
+    region.set_yhi(tmp_box.yMax());
+    lonely_vias_regions.emplace_back(region);
+  }
+  // merge intersecting regions
+  std::sort(lonely_vias_regions.begin(),
+            lonely_vias_regions.end(),
+            [](const Rect& a, const Rect& b) { return a.xMin() < b.xMin(); });
+  std::vector<Rect> merged_regions;
+  for (const auto& region : lonely_vias_regions) {
+    bool found = false;
+    for (auto& merged_region : merged_regions) {
+      if (region.intersects(merged_region)) {
+        merged_region.merge(region);
+        found = true;
+        break;
+      }
+    }
+    if (!found) {
+      merged_regions.emplace_back(region);
+    }
+  }
+  // create drWorkers for the final regions
+  omp_set_num_threads(MAX_THREADS);
+#pragma omp parallel for schedule(dynamic)
+  for (size_t i = 0; i < merged_regions.size(); i++) {
+    auto route_box = merged_regions.at(i);
+    auto worker = std::make_unique<FlexDRWorker>(&via_data_, design_, logger_);
+    Rect ext_box;
+    Rect drc_box;
+    auto minGcellIdx = getDesign()->getTopBlock()->getGCellIdx(
+        {route_box.xMin(), route_box.yMin()});
+    route_box.bloat(MTSAFEDIST, ext_box);
+    route_box.bloat(DRCSAFEDIST, drc_box);
+    worker->setRouteBox(route_box);
+    worker->setExtBox(ext_box);
+    worker->setDrcBox(drc_box);
+    worker->setDRIter(64);
+    worker->setDebugSettings(router_->getDebugSettings());
+    worker->setRipupMode(RipUpMode::VIASWAP);
+    worker->setGraphics(graphics_.get());
+    worker->main(getDesign());
+#pragma omp critical
+    {
+      worker->end(getDesign());
+    }
+  }
+}
+
+std::vector<frVia*> FlexDR::getLonelyVias(frLayer* layer,
+                                          int max_spc,
+                                          int cut_class)
+{
+  std::vector<frVia*> lonely_vias;
+  if (layer->getSecondaryViaDefs().empty()) {
+    return lonely_vias;
+  }
+  auto vias = getRegionQuery()->getVias(layer->getLayerNum());
+  std::vector<std::pair<int, int>> via_positions;
+  via_positions.reserve(vias.size());
+  for (auto [obj, box] : vias) {
+    via_positions.emplace_back(box.xCenter(), box.yCenter());
+  }
+  KDTree tree(via_positions);
+  std::vector<bool> visited(via_positions.size(), false);
+  std::vector<int> isolated_via_nodes;
+  omp_set_num_threads(ord::OpenRoad::openRoad()->getThreadCount());
+#pragma omp parallel for schedule(dynamic)
+  for (int i = 0; i < via_positions.size(); i++) {
+    bool already_visited = false;
+#pragma omp critical
```

Comment:
I'm confused about this code.  There appears to be non-determinism.  If 0 is going to visit 1 as a neighbor then after 0 finishes the first critical section it depends whether 1 enters the first section (unvisited) or 0 enters the second critical section later causes 1 to skip as visited.

Reply:
There are 2 scenarios that could happen:
- 1 enters the first critical section before 0 enters the second critical section: In that case, 1 would be visited twice; once by 0 as a neighbor and the other as the main node to be searched. In terms of memory writes, visited[1] = true; would be called twice, which isn't harmful. isolated_via_nodes would not be touched because this only happens to nodes with neighbors.
- 1 enters the first critical section after 0 exits the second critical section: 1 would have already been marked as visited so it would be skipped.

Parallelization here is done to speedup the search process when possible. it could happen that in some cases the work would be done twice but the result should be the same

Reply:
As two different threads could be read/writing visited I think you need to use `std::vector<atomic_bool> visited(via_positions.size(), false);`

Reply:
Also in omp "All unnamed critical directives map to the same name" so I think you need to name these sections if you want to allow them to run in parallel.  Note names are global.

Reply:
If I understand correctly. both critical sections cannot be executed in parallel so it would never happen that two threads could be read/writing visited at the same time.

> A thread waits at the start of a critical region identified by a given name until no other thread in the program is executing a critical region with that same name. Critical sections not specifically named by **omp critical** directive invocation are mapped to the same unspecified name.

Reply:
With the use of atomics this block doesn't need to be critical

Reply:
Note that omp also has atomics (#pragma omp atomic).   I don't know if there is any advantage to one over the other

Reply:
I just remembered that std::vector<bool> is not thread-safe (we had a similar issue in the connectivity checker).  It uses bit packing so the individual elements are not truly separate.  I had to change it to vector<char>